### PR TITLE
fix(transaction): stops dead locking on multiple concurrent transactions

### LINKED ISSUE #11024: Deadlock on multiple transactions
## What are you doing?

```js
const Sequelize = require('sequelize');
const sequelize = new Sequelize('mysql://user:password@dburl:3306/test')

class TestUser extends Sequelize.Model {}
TestUser.init({
  username: Sequelize.STRING,
  birthday: Sequelize.DATE
}, { sequelize, modelName: 'testuser' })

for (var i = 0; i < 5; i++) {
    sequelize.transaction(t => {
        return TestUser.findOne({
            where: { username: 'janedoe' }
        }, { transaction: t })
        .then(user => {
            return TestUser.update(
                { birthday: new Date(1980, 1, 8) },
                {
                    where: { username: 'johndoe' }
                },
                { transaction: t }
            )
        })
    })
    .then(() => console.log('done'))
    .catch(err => console.log(err))
}
```

The loop is to mimic consecutive requests to a server. Adjusting the loop count below the connection pool max (default is 5) stops the problem.

**To Reproduce**
1. Run the quick example in the documentation (to create an entry)
2. Run the code above
3. See all of updates above fail

## What do you expect to happen?
Update the entry successfully.

## What is actually happening?
```
Executing (acafd742-e1e0-4181-be35-ded526d3d4d6): START TRANSACTION;
Executing (358028a0-3206-4262-b96a-f06f9359ecb8): START TRANSACTION;
Executing (eb448c6b-2098-499c-8d60-b6b4c83b2c23): START TRANSACTION;
Executing (cd0d76fc-78a0-4799-bcce-82451fcc38fd): START TRANSACTION;
Executing (23e9b61e-91ab-4764-85a3-98893e0d0f5a): START TRANSACTION;
Executing (acafd742-e1e0-4181-be35-ded526d3d4d6): ROLLBACK;
Executing (358028a0-3206-4262-b96a-f06f9359ecb8): ROLLBACK;
Executing (eb448c6b-2098-499c-8d60-b6b4c83b2c23): ROLLBACK;
Executing (cd0d76fc-78a0-4799-bcce-82451fcc38fd): ROLLBACK;
Executing (23e9b61e-91ab-4764-85a3-98893e0d0f5a): ROLLBACK;
{ SequelizeConnectionAcquireTimeoutError: Operation timeout
...
```

## Environment
Dialect:
- [x] mysql
- [ ] postgres
- [ ] sqlite
- [ ] mssql
- [ ] any
Dialect **library** version: 1.6.5
Database version: 5.7.24
Sequelize version: 5.8.7
Node Version: 8.12.0
OS: Linux 4.19.23-gentoo (x86_64)
If TypeScript related: TypeScript version: N/A
Tested with latest release:
- [ ] No
- [X] Yes, specify that version: 5.8.7

### PR DESCRIPTION
**Pull request by agirorn**
### Pull Request check-list

_Please make sure to review and check all of these items:_

- [x] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)?
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Have you added new tests to prevent regressions?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [ ] Did you update the typescript typings accordingly (if applicable)?
- [x] Did you follow the commit message conventions explained in [CONTRIBUTING.md](https://github.com/sequelize/sequelize/blob/master/CONTRIBUTING.md)?

### Description of change

Adding a queue that limits the number of currently running transactions
to equal the maxSize of the configured connection pool.

closes #11024

### CODE REVIEW

File: lib/sequelize.js [hunk 1]
```diff
@@ -22,6 +22,7 @@ const Association = require('./associations/index');
 const Validator = require('./utils/validator-extras').validator;
 const Op = require('./operators');
 const deprecations = require('./utils/deprecations');
+const { default: PQueue } = require('p-queue');
```
Comment (papb) [inline lib/sequelize.js]: Minor refactor (code style preference)

```suggestion
const PQueue = require('p-queue').default;
```

File: lib/sequelize.js [hunk 2]
```diff
@@ -1100,7 +1107,7 @@ class Sequelize {
     if (!autoCallback) return transaction.prepareEnvironment(false).return(transaction);
 
     // autoCallback provided
-    return Sequelize._clsRun(() => {
+    return Sequelize._clsRun(this.queued(() => {
```
Comment (papb) [inline lib/sequelize.js]: Interesting, from this it seems that you're only fixing this behavior for CLS, does it really fix it for manual transactions as well?

Comment (agirorn) [reply lib/sequelize.js]: I just don't know, but I will look into it that.

- [ ] Look into dead-lock in manual transactions

**Comment by papb** [comment]
Hello @agirorn, are you still up to this PR?

**Comment by agirorn** [comment]
Sorry for the late replay.

Not I do not have the capacity for this work. And also do not know exactly how I'd accomplish this for all the other cases.

File: test/integration/transaction.test.js [hunk 3]
```diff
@@ -366,6 +366,41 @@ if (current.dialect.supports.transactions) {
       });
     });
 
+    describe('more concurrent transactios than the connections in connection pool', () => {
```
Comment (DannyNemer) [inline test/integration/transaction.test.js:369]:
```suggestion
    describe('more concurrent transactions than the connections in connection pool', () => {
```

**Comment by perrosnk** [comment]
I am also having the same issue. Is there any timeline for when this will be merged?

**Comment by agirorn** [comment]
Closing this PR since I do not have the capacity to complete it and no longer use sequelize.

**Pull request closed without merging**